src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -284,15 +301,31 @@ private bool CData()
             }
             else
             {
-                var tagName = String.Empty;
-                if (At(HtmlSymbolType.Text))
+                var tagName = string.Empty;
+                HtmlSymbol bangSymbol = null;
+
+                if (At(HtmlSymbolType.Bang))
+                {
+                    bangSymbol = CurrentSymbol;
+
+                    var nextSymbol = Lookahead();
+
+                    if (nextSymbol.Type == HtmlSymbolType.Text)
+                    {
+                        tagName = nextSymbol.Content;
+                    }
+                }
+                else if (At(HtmlSymbolType.Text))
                 {
                     tagName = CurrentSymbol.Content;
                 }
+
                 var matched = RemoveTag(tags, tagName, tagStart);
 
                 if (tags.Count == 0 &&
-                    String.Equals(tagName, SyntaxConstants.TextTagName, StringComparison.OrdinalIgnoreCase) &&
+                    // text tags cannot be targeted by TagHelpers and therefore can't be escaped with '!'.
```

Comment:
Whaaa, a `<text>` tag doesn't look anything like a `<p>` as far as syntax trees go, and due to that limitation is incapable to be targeted by TagHelpers. Same principle applies to `<![CDATA[..`. If someone happens to write `!` we want to treat it as a normal HTML tag, that's what this check ensures will happen.